crates/net/net.rs

```diff
@@ -49,53 +52,166 @@ async fn discover_peers(udp_addr: SocketAddr, signer: SigningKey, bootnodes: Vec
+                    if peer.last_ping_hash.is_none() {
+                        warn!("Discarding pong as the node did not send a previous ping");
+                        continue;
+                    }
+                    if peer.last_ping_hash.unwrap() == msg.ping_hash {
+                        peer.last_ping_hash = None;
+                        peer.is_proven = true;
+                    } else {
+                        warn!(
+                            "Discarding pong as the hash did not match the last corresponding ping"
+                        );
+                    }
+                } else {
+                    warn!("Discarding pong as it is not a known node");
+                }
+            }
+            Message::FindNode(msg) => {
+                if is_expired(msg.expiration) {
+                    warn!("Ignoring find node msg as it is expired.");
+                    continue;
+                };
+                let node = table.get_by_node_id(packet.get_node_id());
+                if let Some(node) = node {
+                    if node.is_proven {
+                        let nodes = table.get_closest_nodes(node.node.node_id);
+                        let expiration = get_expiration(20);
+                        let neighbors =
+                            discv4::Message::Neighbors(NeighborsMessage::new(nodes, expiration));
+                        let mut buf = Vec::new();
+                        neighbors.encode_with_header(&mut buf, &signer);
+                        info!("Sending neighbors!");
+                        udp_socket.send_to(&buf, from).await.unwrap();
+                    } else {
+                        warn!("Ignoring find node message as the node isn't proven!");
+                    }
+                } else {
+                    warn!("Ignoring find node message as it is not a known node");
+                }
             }
             Message::Neighbors(neighbors_msg) => {
-                let nodes = &neighbors_msg.nodes;
-                for node in nodes {
-                    let peer_data = PeerData::from(*node);
-                    table.insert(peer_data);
-                    let node_addr = SocketAddr::new(node.ip.to_canonical(), node.udp_port);
-                    ping(&udp_socket, udp_addr, node_addr, &signer).await;
+                if is_expired(neighbors_msg.expiration) {
+                    warn!("Ignoring neighbor msg as it is expired.");
+                    continue;
+                };
+
+                let mut nodes_to_insert = None;
+                if let Some(node) = table.get_by_node_id_mut(packet.get_node_id()) {
+                    if let Some(req) = &mut node.find_node_request {
+                        let nodes = &neighbors_msg.nodes;
+                        let nodes_sent = req.nodes_sent + nodes.len();
+
+                        if nodes_sent <= MAX_NODES_PER_BUCKET {
+                            info!("Storing neighbors in our table!");
```

Comment:
I would change all these new `info` to `debug` to avoid adding noise to the logs